ci/generate_docs.sc

```diff
@@ -0,0 +1,150 @@
+#!/usr/bin/env amm
+
+import java.time.Instant
+import ammonite.ops._
+import ammonite.ops.ImplicitWd._
+
+import $file.utils
+import utils.SemVer
+
+
+/**
+  * Docs generation process:
+  *
+  * 1. copy docs/docs and docs/_layouts/docs.html part from 1.3, 1.4 and 1.5 branches to the temp dirs
+  * 2. copy the latest release 1.5 branch doc to the temp dir
+  * 3. generate the latest release docs
+  * 4. replace docs/docs with docs/docs from 1.3, 1.4, 1.5 and generate docs for the _sites/docs_version
+  */
+
+val ignoredReleaseBranchesVersions = Set("1.6")
+
+val currentGitBranch = %%('git, "rev-parse", "--abbrev-ref", "HEAD").out.lines.head
+
+@main
+def main(tracked_by_git: Boolean = true) = {
+  if (tracked_by_git) {
+    // If the script is being tracked by git, we should create and run an untracked copy of it in order
+    // to avoid errors during branch checkout
+    cp.over(pwd/"generate_docs.sc", pwd/"generate_docs.temp")
+    %('amm, "generate_docs.temp", "--tracked_by_git", "false")
+    rm! pwd/"generate_docs.temp"
+  } else {
+    // check that repo is clean
+    val possiblyChangedLines: Vector[String] = %%('git, "diff-files", "--name-only").out.lines
+    if (possiblyChangedLines.nonEmpty) {
+      val msg = s"Git repository isn't clean, aborting docs generation. Changed files:\n${possiblyChangedLines.mkString("\n")}"
+      println(msg)
+      exit()
+    }
+
+    val docsBuildDir = makeTmpDir()
+    val docsSourceDir = pwd/up/"docs"
+
+    buildDocs(docsBuildDir, docsSourceDir)
+    %('git, 'checkout, currentGitBranch)
+    println(s"Success! Docs were generated at ${docsBuildDir/'docs/'_site}")
+  }
+}
+
+def makeTmpDir(): Path = {
+  val timestamp = Instant.now().toString.replace(':', '-')
+  val path = root/"tmp"/s"marathon-docs-build-$timestamp"
+  mkdir! path
+  path
+}
+
+
+
+def listAllTagsInOrder: Vector[String] = %%('git, "tag", "-l", "--sort=version:refname").out.lines
+
+def isReleaseTag(tag: String) = tag.matches("""v[1-9]+\.\d+\.\d+""")
+
+/**
+  * get the major.minor version from a [v]major.minor.patch
+  *
+  * v1.4.11 -> 1.4
+  * 1.45.983 -> 1.45
+  *
+  */
+def toSemanticVersion(tag: String): SemVer = {
+  val verionOnly = tag.filterNot(c => c.isLetter || c == '-')
+  val weDontUseCommitSha = ""
+  SemVer(verionOnly, weDontUseCommitSha)
+}
+
+def notIgnoredBranch(branchVersion: String) = !ignoredReleaseBranchesVersions.contains(branchVersion)
+
+def getLatestPatch(tags: Seq[SemVer]): SemVer = {
+  tags.last
+}
+
+/**
+  * getting 3 latest release versions and corresponding latest tags
+  *
+  * example: List[(String, String)] = List(("1.3", "1.3.14"), ("1.4", "1.4.11"), ("1.5", "1.5.6"))
+  */
+val docsTargetVersions: List[(String, SemVer)] = listAllTagsInOrder.filter(isReleaseTag).map(toSemanticVersion).groupBy(version => s"${version.major}.${version.minor}").filterKeys(notIgnoredBranch).mapValues(getLatestPatch).toList.sortBy(_._1).takeRight(3)
```

Comment:
Seems like maybe we could add a method, dropRight, to SemVer?

Reply:
It's not easy to follow this. So we just want to know the highest patch version, right? We can always get "major.minor".

```scala
val docsTargetVersions: List[SemVer] = {
   val minors: Map[(Int, Int), List[SemVer]] = listAllTagsInOrder
      .filter(isReleaseTag)
      .map(toSemanticVersion)
      .groupBy(v => (v.major, v.minor))

    minors.values.map(_.max)
}
```